### PR TITLE
crimson: add support for ms_bind_retry_{delay,count}

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/net/Socket.h [hunk 1]
```diff
@@ -160,6 +160,11 @@ class Socket
   friend class FixedCPUServerSocket;
 };
 
+using listen_ertr = crimson::errorator<
+  crimson::ct_error::address_in_use, // The address is already bound
+  crimson::ct_error::address_not_available // https://techoverflow.net/2021/08/06/how-i-fixed-python-oserror-errno-99-cannot-assign-requested-address/The address is already bound
```
Comment (cyx1231st) [inline src/crimson/net/Socket.h:165]: Remove the trailing "The address is already bound"

File: src/crimson/net/SocketMessenger.cc [hunk 2]
```diff
@@ -102,27 +102,29 @@ SocketMessenger::try_bind(const entity_addrvec_t& addrs,
...
-    }).then([] (bool success) -> bind_ertr::future<> {
-      if (success) {
-        return bind_ertr::now();
-      } else {
+    }).then([] (const std::error_code e) -> bind_ertr::future<> {
+      if (!e) {
+        return bind_ertr::now(); // success!
+      } else if (e == std::errc::address_in_use) {
         return crimson::ct_error::address_in_use::make();
+      } else if (e == std::errc::address_in_use) {
```
Comment (cyx1231st) [inline src/crimson/net/SocketMessenger.cc:118]: `std::errc::address_not_available`?

File: src/crimson/net/SocketMessenger.cc [hunk 3]
```diff
@@ -126,6 +118,53 @@ SocketMessenger::try_bind(const entity_addrvec_t& addrs,
...
+          return seastar::make_ready_future<std::optional<std::error_code>>(
+            std::make_optional<std::error_code>(error));
+        }
+      }));
+    }).then([this] (const std::error_code error) -> bind_ertr::future<> {
+      if (!error) {
+        return bind_ertr::now(); // success!
+      } else if (error == std::errc::address_in_use) {
+        return crimson::ct_error::address_in_use::make();
+      } else if (error == std::errc::address_in_use) {
```
Comment (cyx1231st) [inline src/crimson/net/SocketMessenger.cc:164]: `std::errc::address_not_available`?

**Review by cyx1231st — COMMENTED** [review]
Make check does not pass.

**Comment by rzarzynski** [comment]
Thanks for the review, @cyx1231st! Addressed the comments. The build problem was:

```
../src/crimson/net/SocketMessenger.cc:159:14: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
    }).then([this] (const std::error_code error) -> bind_ertr::future<> {
```

So `-Werror` & Clang are doing their work! :-)

**Comment by cyx1231st** [comment]
`../src/test/crimson/test_socket.cc:78:12: error: no member named 'listen_ertr' in 'crimson::net::FixedCPUServerSocket'; did you mean 'crimson::net::listen_ertr'?`

**Comment by rzarzynski** [comment]
@cyx1231st: fixed the unit test.

**Review by cyx1231st — APPROVED** [review]
LGTM!

**Pull request merged**